### PR TITLE
Table change check

### PR DESCRIPTION
**Pull request by twuebi**
#58

### CODE REVIEW

File: crates/iceberg-catalog/src/catalog/tables.rs [hunk 1]
```diff
@@ -819,6 +842,19 @@ impl<C: Catalog, A: AuthZHandler, S: SecretStore>
...
+        for ((update, (_, table_uuid)), response) in updates
+            .into_iter()
+            .zip(&event_table_ids)
+            .zip(&commit_response)
+        {
+            state
+                .v1_state
+                .contract_verifiers
+                .check(&update, *table_uuid, &response.previous_table_metadata)
+                .await?
```
Comment (c-thiel) [inline crates/iceberg-catalog/src/catalog/tables.rs:854]: There might be quite a few updates here - I would prefer concurrent execution and try_join

File: crates/iceberg-catalog/src/catalog/tables.rs [hunk 2]
```diff
@@ -819,6 +842,19 @@ impl<C: Catalog, A: AuthZHandler, S: SecretStore>
             ));
         }
 
+        for ((update, (_, table_uuid)), response) in updates
```
Comment (c-thiel) [inline crates/iceberg-catalog/src/catalog/tables.rs:845]: I think we should push the whole block up to right after C::commit_table_transaction.
There is no need to load storage secret if the contract check fails.

File: crates/iceberg-catalog/src/catalog/tables.rs (hunk 1 above)
Comment (c-thiel) [reply crates/iceberg-catalog/src/catalog/tables.rs:854]: It doesn't quite fit here, but I am also a bit hesitant about awaiting the emit_change_event block.
We currently implement a "send and forget" approach. But right now, if the event queue is offline, we are blocking the client response until the event queue gets a timeout.
What's your feeling about emitting events non-blocking?

File: crates/iceberg-catalog/src/catalog/tables.rs [hunk 3]
```diff
@@ -929,3 +965,65 @@ fn maybe_body_to_json(request: impl Serialize) -> serde_json::Value {
         serde_json::Value::Null
     }
 }
+
+// TableUpdate is not clone but all containing fields are so we use this function to clone..
+fn copy_updates(updates: &[TableUpdate]) -> Vec<TableUpdate> {
```
Comment (c-thiel) [inline crates/iceberg-catalog/src/catalog/tables.rs:970]: Should we maybe rename to clone_updates?

https://github.com/apache/iceberg-rust/pull/402

File: crates/iceberg-catalog/src/service/contract_verification.rs [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+#[async_trait]
+impl ContractVerification for ContractVerifiers {
+    fn name(&self) -> &'static str {
+        "ContractVerifiers"
+    }
+
+    async fn check(
+        &self,
+        table_updates: &[TableUpdate],
+        table_ident_uuid: TableIdentUuid,
```
Comment (c-thiel) [inline crates/iceberg-catalog/src/service/contract_verification.rs:190]: Now that we send the ` current_metadata`, we might want to skip the table_ident_uuid.
The table-uuid is already part of `TableMetadata` and might change during updates.
If we  specify it here additionally, we should rename it to current_table_ident_uuid to make sure it is the UUID before updates.
I think removing it is fine though.

File: crates/iceberg-catalog/src/service/contract_verification.rs [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    ) -> Result<ContractVerificationOutcome, ErrorModel> {
+        for checker in &self.checkers {
+            match checker
+                .check(table_updates, table_ident_uuid, current_metadata)
+                .await
+            {
+                Ok(ContractVerificationOutcome::Clear {}) => {}
+                Ok(block_result @ ContractVerificationOutcome::Violation { error_model: _ }) => {
+                    tracing::info!(
+                        "Checker {} blocked change on table '{}'",
```
Comment (c-thiel) [inline crates/iceberg-catalog/src/service/contract_verification.rs:201]: Lets reflect the trait in the log - maybe ContractValidator {} blocked change on table ..
Also for other logs here.

File: crates/iceberg-catalog/src/catalog/tables.rs (hunk 1 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/catalog/tables.rs:901]: Do we expect this endpoint to be called a lot? How many changes do we expect per call? 

We could separate it out via a channel and have some background task shoveling it into nats.

File: crates/iceberg-catalog/src/service/contract_verification.rs (hunk 5 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/service/contract_verification.rs]: done

File: crates/iceberg-catalog/src/catalog/tables.rs (hunk 3 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/catalog/tables.rs]: done

File: crates/iceberg-catalog/src/service/contract_verification.rs (hunk 4 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/service/contract_verification.rs]: thanks, missed it in table-metadata, removed!

File: crates/iceberg-catalog/src/catalog/tables.rs (hunk 2 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/catalog/tables.rs]: done

File: crates/iceberg-catalog/src/catalog/tables.rs (hunk 1 above)
Comment (twuebi) [reply crates/iceberg-catalog/src/catalog/tables.rs]: > There might be quite a few updates here - I would prefer concurrent execution and try_join

made it futures + `try_join_all`

Comment (twuebi) [reply crates/iceberg-catalog/src/catalog/tables.rs]: > Do we expect this endpoint to be called a lot? How many changes do we expect per call?
> 
> We could separate it out via a channel and have some background task shoveling it into nats.

https://github.com/hansetag/iceberg-catalog/issues/86

**Review by c-thiel — APPROVED** [review]

**Pull request merged**